apps/trivium/src/trivium/trivium_byte.rs

```diff
@@ -0,0 +1,258 @@
+            self.next_64();
+        }
+    }
+
+    /// Computes 8 potential future step of Trivium, b*8 terms in the future. This does not update
+    /// registers, but rather returns with the output, the three values that will be used to
+    /// update the registers, when the time is right. This function is meant to be used in
+    /// parallel.
+    fn get_output_and_values(
+        a_byte: StaticBitDeque<12, T>,
+        b_byte: StaticBitDeque<11, T>,
+        c_byte: StaticBitDeque<14, T>,
+        b: usize,
+    ) -> [T; 4] {
+        let n = b * 8 + 7;
+        assert!(n < 65);
+
+        let ((a1, a2, a3, a4, a5), ((b1, b2, b3, b4, b5), (c1, c2, c3, c4, c5))) = rayon::join(
+            || Self::get_bytes(a_byte, [91 - n, 90 - n, 68 - n, 65 - n, 92 - n]),
+            || {
+                rayon::join(
+                    || Self::get_bytes(b_byte, [82 - n, 81 - n, 77 - n, 68 - n, 83 - n]),
+                    || Self::get_bytes(c_byte, [109 - n, 108 - n, 86 - n, 65 - n, 110 - n]),
+                )
+            },
+        );
+
+        let (((temp_a, temp_b), (temp_c, a_and)), (b_and, c_and)) = rayon::join(
+            || {
+                rayon::join(
+                    || rayon::join(|| a4 ^ a5, || b4 ^ b5),
+                    || rayon::join(|| c4 ^ c5, || a1 & a2),
+                )
+            },
+            || rayon::join(|| b1 & b2, || c1 & c2),
+        );
+
+        let (temp_a_2, temp_b_2, temp_c_2) = (temp_a.clone(), temp_b.clone(), temp_c.clone());
+
+        let ((o, a), (b, c)) = rayon::join(
+            || {
+                rayon::join(
+                    || (temp_a_2 ^ temp_b_2) ^ temp_c_2,
+                    || temp_c ^ ((c_and) ^ a3),
+                )
+            },
+            || rayon::join(|| temp_a ^ (a_and ^ b3), || temp_b ^ (b_and ^ c3)),
+        );
+
+        [o, a, b, c]
+    }
+
+    /// This calls `get_output_and_values` in parallel 8 times, and stores all results in a Vec.
+    fn get_64_output_and_values(&self) -> Vec<[T; 4]> {
+        let copies = vec![
+            (
+                self.a_byte.clone(),
+                self.b_byte.clone(),
+                self.c_byte.clone()
+            );
+            8
+        ];
+        copies
+            .into_par_iter()
+            .enumerate()
+            .map(|(i, (a, b, c))| Self::get_output_and_values(a, b, c, i))
+            .collect()
+    }
+
+    /// Computes 64 turns of the stream, outputting the 64 bits (in 8 bytes) all at once in a
+    /// Vec (first value is oldest, last is newest)
+    pub fn next_64(&mut self) -> Vec<T> {
+        match &self.fhe_key {
+            Some(sk) => {
+                rayon::broadcast(|_| set_server_key(sk.clone()));
+            }
+            None => (),
+        }
+        let values = self.get_64_output_and_values();
+        match &self.fhe_key {
+            Some(_) => {
+                rayon::broadcast(|_| unset_server_key());
+            }
+            None => (),
+        }
```

Comment:
is it necessary to unset ? can't we assume computations will set their keys if required ?

The only usage where I think this could make sense is with a "stack" of server keys where we restore the previous key that was there, in that case I don't think I see the point of unsetting our key ? unless it allows to drop memory more easily, but I guess we don't have an actual copy of the BSK in each thread right ?

It's a general remark, I won't pin point all unset sk calls

Reply:
My thinking was that a potential user of this implementation might have other things going on in their life (other computations with other server keys). I didn't want for the implem to have as a side effect that every thread of the rayon pool will have this key set. (Imagine if they forget to set their server key: I'd rather see them have a runtime error than incorrect computations)

There is one alternative: that the user is responsible for setting the key before calling this. But with rayon/multithreading this is clunky because they will have to use rayon::broadcast or equivalent.

This seemed (and seems) best to me for now